### ISSUE TITLE
Test

### ISSUE THREAD

**Issue by lmackenzie94** [lmackenzie94/eleventy-practice#2]
Signs it's time to consider extracting a new component
Duplicated JSX (extract to the same file if not reused elsewhere)
Reuse
JSX gets difficult to read
To allow thinking about page sections in isolation (each component becomes a black box, which reduces cognitive load)
Props list gets long
Document different states in Storybook, create dedicated image tests with Chromatic/Percy/etc
API calls and JSX in same component - Extracting API concerns means you can feed mock data via props
Performance (extract an expensive portion so you can optimize its rendering)
To keep style selectors short (whether it be BEM or CSS Modules, smaller components = smaller namespace = can use nice short names and easily navigate). If you find yourself prefixing selectors by section like this, consider extracting a component.
To divide work between people/teams
Why prefer fewer components?
Less jumping between files
Minimize prop drilling and propType declaration overhead
Shallow trees may be easier to understand than a deeply nested tree
Every new file adds a little bloat to the bundle due to Webpack's bundling overhead
So it's a tradeoff.

Useful middle ground here: You can declare multiple components in a single file, but export only the "parent". This avoids jumping between files and provides a clear public API. If you prefer separate files, you can create a component folder and place all its child components in that folder, then use a barrel to export only the Header.

Brain Teaser
Why doesn't the onSubmit fire here?

Patterns for scaling
Useful for large apps, multiple apps in the org, or cross-cutting concerns

 "Vertical" slicing: Split app by top-level nav
 "Tiny" slices / mixed tech: Micro frontends
 Overview on Martin Fowler
 MFE with React and Vue
 Use context in subtrees (see below for context)
 Share component across separate teams
 Create a company framework
 Review large open source projects <- See "Code worth reading" and identify patterns used for scaling
Example Custom Hooks
 useMouseCoordinates - Mouse coordinate tracker
 useApiCall - Wrap API calls in a hook to encapsulate the redundancy of tracking response, loading, and error state. Enhancement: mimic the "Only one loader" behavior by incrementing a counter each time a call starts and decrementing a counter each time a call completes or errors. Share numApiCallsInProgress or an array of apiCallsInProgress with an identifier for each between hooks via context.
 useLocalStorage -
 useAsync - for handling an async call with loading states and errors. blog post
 useForm - simple hook for handling form state and errors
 useRequiredFields - Form with uncontrolled inputs, but values are required - Codesandbox
 Note: Avoid declaring dep array for custom hooks. Expect consumers to call useCallback/useMemo if needed.
 Support both Hooks and render props via the hydra pattern
JSX in Depth
 Spread attributes
 Booleans, null, undefined are ignored
 Children
 Passing props to children
 Demo: Create Card component that displays text inside a box with an optional title. Can also show "slots" pattern.
 Pass component as prop (aka slots)
Key Advanced Features
 Context - For "global" data.
 i18n, user data, theme, auth
 It's been around since the creation of React (and used by React Router and Redux), but not stable until 16.
 Context does NOT have to be global to the whole app.
 You can have multiple logically separated contexts in your app.
 Wrap provider and consumer to enforce certain usage and enhance performance and here
 Optimize your context value if perf is an issue
 Redux remains useful. Redux offers scalable unidirectional flows, decomposed state handling via dedicated reducers and actions, hot reloading, time travel, and potential performance benefits via selectors and connect's reference comparison. Though can consider react-waterfall to get most these benefits (albeit without Redux's huge ecosystem and excellent docs).
 Perf tip: As usual, avoid declaring objects in render since they're recreated on every render which causes needless rerenders. Instead, declare the object passed down via context in state. More here under "Issues with the New API". Can put methods in state too. Or, you can memoize the function
 You can require a parent provider with a simple pattern too.
 3 ways to consume:
 useHook (prefer)
 Consumer via render prop / child as func
 Static contextType (classes only, and can only consume one context this way unless you wrap the others. Bottom line, prefer useHook)
 Connected components re-render when any values passed to the provider's value prop change. So don't declare the value in render. Instead, lift it to state.
 Require context
 Updating context from the consumer - However, consider passing dispatch and using useReducer instead since that may be preferable.
 Consuming multiple contexts
 Create dedicated "store" component that holds provider
 Refs
 React docs
 Prone to abuse. Avoid - you may not need it. Can be misused to manually change the DOM.
 2 common valid uses:
 1. Read and change DOM: focus, and calculate size and position.
 2. Hold values not used in render in func components (can use instance vars in classes).
 Detailed blog post "The Complete Guide to React Refs"
Basically a special case of useState that doesn't trigger a re-render
Set its value directly (instead of using a setter)
Get value via .current property
4 approaches - cheatsheet:
 String refs (deprecated)
 Callback refs (clunkier syntax than createRef, but can be preferable when dealing with dynamic refs - otherwise prefer createRef or useRef below)
 createRef (Classes only - simpler syntax than callbacks)
 useRef (Funcs only) UseRef is basically this
 Render outside the parent's DOM via Portals - Simple example
 Capture scroll position before render using getSnapshotBeforeUpdate. snapshot is the third param passed to componentDidUpdate.
 Strict mode - Example
